pkg/operator/controller/externaldns/deployment.go

```diff
@@ -0,0 +1,274 @@
+	replicas := int32(1)
+
+	matchLbl := map[string]string{
+		appNameLabel:     controller.ExternalDNSBaseName,
+		appInstanceLabel: externalDNS.Name,
+	}
+
+	nodeSelectorLbl := map[string]string{
+		osLabel:             linuxOS,
+		masterNodeRoleLabel: "",
+	}
+
+	depl := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      controller.ExternalDNSResourceName(externalDNS),
+			Namespace: namespace,
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: matchLbl,
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: matchLbl,
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: serviceAccount.Name,
+					NodeSelector:       nodeSelectorLbl,
+				},
+			},
+		},
+	}
+
+	for i, zone := range externalDNS.Spec.Zones {
+		container, err := populateExternalDNSContainer(image, metricsStartPort+i, zone, externalDNS)
+		if err != nil {
+			return nil, err
+		}
+		depl.Spec.Template.Spec.Containers = append(depl.Spec.Template.Spec.Containers, *container)
+	}
+
+	return depl, nil
+}
+
+// WIP function for generating container specs for one container at a time.
+func populateExternalDNSContainer(image string, metricsPort int, zone string, externalDNS *operatorv1alpha1.ExternalDNS) (*corev1.Container, error) {
+	name := "externaldns-" + zone
+
+	args := []string{
+		fmt.Sprintf("--metrics-address=127.0.0.1:%d", metricsPort),
+		fmt.Sprintf("--txt-owner-id=externaldns-%s", externalDNS.Name),
+		fmt.Sprintf("--zone-id-filter=%s", zone),
+		"--policy=sync",
+		"--registry=txt",
+		"--log-level=debug",
+	}
+
+	provider, ok := providerStringTable[externalDNS.Spec.Provider.Type]
+	if !ok {
+		return nil, fmt.Errorf("%s is not a supported provider", externalDNS.Spec.Provider.Type)
+	}
+	args = append(args, fmt.Sprintf("--provider=%s", provider))
+
+	//TODO: Add provider credentials logic
+
+	source, ok := sourceStringTable[externalDNS.Spec.Source.Type]
+	if !ok {
+		return nil, fmt.Errorf("%s is not a supported source type", externalDNS.Spec.Source.Type)
+	}
+	args = append(args, fmt.Sprintf("--source=%s", source))
+
+	if externalDNS.Spec.Source.Namespace != nil && len(*externalDNS.Spec.Source.Namespace) > 0 {
+		args = append(args, fmt.Sprintf("--namespace=%s", *externalDNS.Spec.Source.Namespace))
+	}
+
+	if externalDNS.Spec.Source.Service != nil && len(externalDNS.Spec.Source.Service.ServiceType) > 0 {
+		serviceTypeFilter, publishInternal := "", false
+		for _, serviceType := range externalDNS.Spec.Source.Service.ServiceType {
+			serviceTypeFilter += string(serviceType) + ","
+			if serviceType == corev1.ServiceTypeClusterIP {
+				publishInternal = true
+			}
+		}
+
+		// avoid having a trailing comma
+		args = append(args, fmt.Sprintf("--service-type-filter=%s", serviceTypeFilter[0:len(serviceTypeFilter)-1]))
+
+		if publishInternal {
+			args = append(args, "--publish-internal-services")
```

Comment:
Any idea why `--publish-internal-services` is a separate flag from `--service-type-filter`?  Is it a legacy flag, or is publishing ClusterIP services dangerous or something such that it calls for a separate flag?

Reply:
They preferred this to be an opt-in flag: https://github.com/kubernetes-sigs/external-dns/pull/278

Reply:
So it looks like it's a flag to say, "Yes, I really did mean to include ClusterIP services."  Do we need the same kind of "Yes, I really mean it" flag in the CRD?  I'd rather not, but the fact that upstream added the flag makes me wary.

Reply:
No, didn't see anything like this for CRD. There are some other guard flags for headless services but that not quite the same case as for `ClusterIP` as headless is not a service type so you cannot set in `--service-type-filter`.
```bash
  --publish-host-ip              Allow external-dns to publish host-ip for 
                                 headless services (optional)
  --always-publish-not-ready-addresses  
                                 Always publish also not ready addresses for 
                                 headless services (optional)
```

Reply:
Looked at this thing one more time. Actually you were right about the legacy.

At the time when `--publish-internal-services` was added there was no `--service-type-filter` flag. There was no way to fine tune deeper than the source level, so you could only say: I want services as source (`--source=service`) and yes I want internal ClusterIPs to be published too (`--publish-internal-services`).

Then they added  `--service-type-filter` flag in [PR#653](https://github.com/kubernetes-sigs/external-dns/pull/653) which defaulted to `all` which kept the old combination of `--source=service --publish-internal-services` still relevant. However if the filter is set to `ClusterIP` you have to specify the unnecessary `--publish-internal-services`, otherwise the IP [will not be found](https://github.com/kubernetes-sigs/external-dns/blob/master/source/service.go#L461).